File: sdk/ts/src/llm.ts
/**
 * @packageDocumentation
 *
 * LLM client facade — mirrors Go `llm` package interfaces as TS types.
 *
 * Provides URI parsing, config loading, provider registry, Client facade
 * with fallback, and typed errors. Adapter implementations (OpenAI,
 * Anthropic, Ollama) are deferred; register them via {@link register}.
 *
 * @example
 * ```ts
 * import { createLLM, register } from '@hop-top/kit/llm';
 *
 * register('mock', (cfg) => ({ complete: async (r) => ({ content: 'hi', role: 'assistant' }), close() {} }));
 * const client = createLLM('mock://model-1');
 * const resp = await client.complete({ messages: [{ role: 'user', content: 'hello' }] });
 * ```
 */

// ─── Data types ──────────────────────────────────────────────────────────────

export interface Message {
  role: string;
  content: string;
}

export interface Request {
  messages: Message[];
  model?: string;
  temperature?: number;
  maxTokens?: number;
  stopSequences?: string[];
  extensions?: Record<string, unknown>;
}

export interface Response {
  content: string;
  role: string;
  usage?: Usage;
  finishReason?: string;
}

export interface Usage {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
}

export interface Token {
  content: string;
  done: boolean;
}

export interface ToolDef {
  name: string;
  description: string;
  parameters: unknown;
}

export interface ToolCall {
  id: string;
  name: string;
  arguments: unknown;
}

export interface ToolResponse {
  content: string;
  toolCalls: ToolCall[];
}

// ─── Core interfaces ─────────────────────────────────────────────────────────

export interface Completer {
  complete(req: Request): Promise<Response>;
}

export interface Streamer {
  stream(req: Request): AsyncIterable<Token>;
}

export interface ToolCaller {
  callWithTools(req: Request, tools: ToolDef[]): Promise<ToolResponse>;
}

export interface Provider
  extends Partial<Completer & Streamer & ToolCaller> {
  close(): void;
}

// ─── URI + Config ────────────────────────────────────────────────────────────

export interface URI {
  scheme: string;
  model: string;
  host?: string;
  params?: Record<string, string>;
}

export interface ProviderConfig {
  apiKey?: string;
  baseURL?: string;
  model: string;
  params?: Record<string, string>;
  extras?: Record<string, unknown>;
}

export interface ResolvedConfig {
  uri: URI;
  provider: ProviderConfig;
  fallbacks: string[];
}

// ─── Errors ──────────────────────────────────────────────────────────────────

export class LLMError extends Error {
  statusCode?: number;
  code?: string;

  constructor(message: string) {
    super(message);
    this.name = "LLMError";
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class ProviderNotFoundError extends LLMError {
  scheme: string;

  constructor(scheme: string) {
    super(`provider not found: ${scheme}`);
    this.name = "ProviderNotFoundError";
    this.scheme = scheme;
  }
}

export class CapabilityNotSupportedError extends LLMError {
  capability: string;
  provider: string;

  constructor(capability: string, provider: string) {
    super(
      `capability "${capability}" not supported by provider "${provider}"`
    );
    this.name = "CapabilityNotSupportedError";
    this.capability = capability;
    this.provider = provider;
  }
}

export class AuthError extends LLMError {
  provider: string;

  constructor(provider: string) {
    super(`authentication failed for provider: ${provider}`);
    this.name = "AuthError";
    this.provider = provider;
  }
}

export class RateLimitError extends LLMError {
  provider: string;
  retryAfter?: number;

  constructor(provider: string, retryAfter?: number) {
    super(`rate limit exceeded for provider: ${provider}`);
    this.name = "RateLimitError";
    this.provider = provider;
    this.retryAfter = retryAfter;
    this.statusCode = 429;
  }
}

export class ModelError extends LLMError {
  model: string;
  provider: string;

  constructor(model: string, provider: string) {
    super(`model "${model}" not available on provider "${provider}"`);
    this.name = "ModelError";
    this.model = model;
    this.provider = provider;
  }
}

export class FallbackExhaustedError extends LLMError {
  errors: Error[];

  constructor(errors: Error[]) {
    super(
      `all providers exhausted (${errors.length} failures)`
    );
    this.name = "FallbackExhaustedError";
    this.errors = errors;
  }
}

/**
 * Returns true when `err` should trigger a fallback attempt.
 *
 * Fallbackable conditions:
 * - LLMError with statusCode >= 500
 * - RateLimitError (429)
 * - Connection/timeout errors (ECONNREFUSED, ETIMEDOUT, ECONNRESET)
 */
export function isFallbackable(err: Error): boolean {
  if (err instanceof RateLimitError) return true;
  if (err instanceof LLMError) {
    if (err.code && isConnectionError(err.code)) return true;
    if (err.statusCode !== undefined && err.statusCode >= 500) return true;
    return false;
  }
  // Handle NodeJS.ErrnoException (e.g. ECONNREFUSED, ETIMEDOUT)
  const code = (err as { code?: string }).code;
  if (code && isConnectionError(code)) return true;
  return false;
}

const CONNECTION_CODES = new Set([
  "ECONNREFUSED",
  "ETIMEDOUT",
  "ECONNRESET",
  "ENETUNREACH",
  "EHOSTUNREACH",
]);

function isConnectionError(code: string): boolean {
  return CONNECTION_CODES.has(code);
}

// ─── URI parsing ─────────────────────────────────────────────────────────────

/**
 * Parses an LLM URI string into its components.
 *
 * Format: `scheme://[host[:port]/]model[?params]`
 *
 * When the path portion contains a host:port prefix (detected by the
 * presence of a numeric port), the first segment is treated as host.
 * Otherwise the entire path is the model name.
 */
export function parseURI(raw: string): URI {
  const sep = "://";
  const idx = raw.indexOf(sep);
  if (idx < 0) {
    throw new LLMError(`invalid LLM URI (missing "://"): ${raw}`);
  }

  const scheme = raw.slice(0, idx);
  if (!scheme) {
    throw new LLMError(`invalid LLM URI (empty scheme): ${raw}`);
  }

  let rest = raw.slice(idx + sep.length);
  if (!rest) {
    throw new LLMError(`invalid LLM URI (empty model): ${raw}`);
  }

  // Extract query params
  let params: Record<string, string> | undefined;
  const qIdx = rest.indexOf("?");
  if (qIdx >= 0) {
    const qs = rest.slice(qIdx + 1);
    rest = rest.slice(0, qIdx);
    params = {};
    for (const pair of qs.split("&")) {
      const eqIdx = pair.indexOf("=");
      if (eqIdx > 0) {
        params[pair.slice(0, eqIdx)] = decodeURIComponent(
          pair.slice(eqIdx + 1)
        );
      }
    }
  }

  if (!rest) {
    throw new LLMError(`invalid LLM URI (empty model): ${raw}`);
  }

  // Detect host:port pattern.
  // If first segment matches `word:digits`, treat it as host.
  let host: string | undefined;
  let model: string;

  const hostPortMatch = rest.match(/^([^/]+:\d+)\/(.+)$/);
  if (hostPortMatch) {
    host = hostPortMatch[1];
    model = hostPortMatch[2];
  } else {
    model = rest;
  }

  const uri: URI = { scheme, model };
  if (host) uri.host = host;
  if (params && Object.keys(params).length > 0) uri.params = params;
  return uri;
}

// ─── Config loading ──────────────────────────────────────────────────────────

/**
 * Builds a {@link ResolvedConfig} from a URI string + env vars.
 *
 * When `uriStr` is empty/undefined, falls back to `LLM_PROVIDER`
 * env var as a full default URI. If neither is provided, throws.
 *
 * URI query params `api_key` and `base_url` populate provider
 * config. If `uri.host` is set, it derives a baseURL. Env vars
 * `LLM_API_KEY` and `LLM_BASE_URL` override last.
 *
 * Env vars:
 * - `LLM_PROVIDER` — full default URI when no arg provided
 * - `LLM_API_KEY` — overrides provider API key
 * - `LLM_BASE_URL` — overrides provider base URL
 * - `LLM_FALLBACK` — comma-separated fallback URIs
 */
export function loadConfig(uriStr?: string): ResolvedConfig {
  const effective = uriStr || process.env["LLM_PROVIDER"];
  if (!effective) {
    throw new LLMError(
      "no LLM URI provided and LLM_PROVIDER env var is not set"
    );
  }

  const uri = parseURI(effective);

  const fallbackEnv = process.env["LLM_FALLBACK"];
  const fallbacks = fallbackEnv
    ? fallbackEnv.split(",").map((s) => s.trim()).filter(Boolean)
    : [];

  const provider: ProviderConfig = {
    model: uri.model,
  };
  if (uri.params) provider.params = { ...uri.params };

  // Extract api_key and base_url from URI query params
  if (uri.params?.api_key) {
    provider.apiKey = uri.params.api_key;
  }
  if (uri.params?.base_url) {
    provider.baseURL = uri.params.base_url;
  }

  // Derive baseURL from host if present
  if (uri.host) {
    provider.baseURL = `http://${uri.host}`;
  }

  // Env var overrides apply LAST
  const envApiKey = process.env["LLM_API_KEY"] || undefined;
  const envBaseURL = process.env["LLM_BASE_URL"] || undefined;
  if (envApiKey) provider.apiKey = envApiKey;
  if (envBaseURL) provider.baseURL = envBaseURL;

  return { uri, provider, fallbacks };
}

// ─── Registry ────────────────────────────────────────────────────────────────

export type Factory = (cfg: ResolvedConfig) => Provider;

const registry = new Map<string, Factory>();

/** Register a provider factory for the given URI scheme. */
export function register(scheme: string, factory: Factory): void {
  if (registry.has(scheme)) {
    throw new LLMError(`provider already registered for scheme: ${scheme}`);
  }
  registry.set(scheme, factory);
}

/** Resolve a URI to a Provider instance via the registry. */
export function resolve(uriStr: string): Provider {
  const cfg = loadConfig(uriStr);
  const factory = registry.get(cfg.uri.scheme);
  if (!factory) {
    throw new ProviderNotFoundError(cfg.uri.scheme);
  }
  return factory(cfg);
}

/** Clear the registry (for tests). */
export function clearRegistry(): void {
  registry.clear();
}

// ─── Client ──────────────────────────────────────────────────────────────────

export interface ClientOptions {
  fallback?: string[];
  onRequest?: (req: Request) => void;
  onResponse?: (resp: Response, durationMs: number) => void;
  onError?: (err: Error) => void;
  onFallback?: (from: number, to: number, err: Error) => void;
}

export interface Client {
  complete(req: Request): Promise<Response>;
  stream(req: Request): AsyncIterable<Token>;
  callWithTools(req: Request, tools: ToolDef[]): Promise<ToolResponse>;
  capabilities(): string[];
  provider(): Provider;
  close(): void;
}

/**
 * Create an LLM client with fallback support and event hooks.
 *
 * The returned Client delegates to the provider resolved from `uri`.
 * On fallbackable errors, it tries each fallback URI in order.
 */
export function createLLM(uri: string, opts?: ClientOptions): Client {
  const cfg = loadConfig(uri);

  // Merge fallbacks: opts > env > none
  const fallbackURIs = opts?.fallback ?? cfg.fallbacks;

  // Resolve primary; cache fallback providers for close()
  const primary = resolve(uri);
  const allURIs = [uri, ...fallbackURIs];
  const resolvedProviders = new Map<number, Provider>();
  resolvedProviders.set(0, primary);

  function resolveAt(index: number): Provider {
    let p = resolvedProviders.get(index);
    if (!p) {
      p = resolve(allURIs[index]);
      resolvedProviders.set(index, p);
    }
    return p;
  }

  async function withFallback<T>(
    op: string,
    fn: (p: Provider) => Promise<T>
  ): Promise<T> {
    const errors: Error[] = [];
    for (let i = 0; i < allURIs.length; i++) {
      try {
        const p = resolveAt(i);
        if (
          !(op in p) ||
          typeof (p as unknown as Record<string, unknown>)[op] !== "function"
        ) {
          throw new CapabilityNotSupportedError(
            op,
            parseURI(allURIs[i]).scheme
          );
        }
        return await fn(p);
      } catch (err) {
        const e = err instanceof Error ? err : new Error(String(err));
        errors.push(e);
        opts?.onError?.(e);

        // CapabilityNotSupportedError: skip to next provider
        const canContinue =
          isFallbackable(e) || e instanceof CapabilityNotSupportedError;

        if (!canContinue || i === allURIs.length - 1) {
          if (errors.length > 1) {
            throw new FallbackExhaustedError(errors);
          }
          throw e;
        }
        opts?.onFallback?.(i, i + 1, e);
      }
    }
    throw new FallbackExhaustedError(errors);
  }

  return {
    async complete(request: Request): Promise<Response> {
      opts?.onRequest?.(request);
      const start = Date.now();
      const resp = await withFallback("complete", (p) =>
        p.complete!(request)
      );
      opts?.onResponse?.(resp, Date.now() - start);
      return resp;
    },

    stream(request: Request): AsyncIterable<Token> {
      // Route stream through fallback. On creation or first-token
      // failure with a fallbackable/capability error, try next provider.
      return {
        [Symbol.asyncIterator]() {
          let inner: AsyncIterator<Token> | null = null;
          let initialized = false;

          return {
            async next(): Promise<IteratorResult<Token>> {
              if (!initialized) {
                initialized = true;
                opts?.onRequest?.(request);

                const errors: Error[] = [];
                for (let i = 0; i < allURIs.length; i++) {
                  try {
                    const p = resolveAt(i);
                    if (
                      !("stream" in p) ||
                      typeof p.stream !== "function"
                    ) {
                      throw new CapabilityNotSupportedError(
                        "stream",
                        parseURI(allURIs[i]).scheme
                      );
                    }
                    const iterable = p.stream(request);
                    const iter = iterable[Symbol.asyncIterator]();
                    // Attempt first read to detect immediate failures
                    const first = await iter.next();
                    inner = iter;
                    return first;
                  } catch (err) {
                    const e =
                      err instanceof Error
                        ? err
                        : new Error(String(err));
                    errors.push(e);
                    opts?.onError?.(e);
                    const canContinue =
                      isFallbackable(e) ||
                      e instanceof CapabilityNotSupportedError;
                    if (!canContinue || i === allURIs.length - 1) {
                      if (errors.length > 1)
                        throw new FallbackExhaustedError(errors);
                      throw e;
                    }
                    opts?.onFallback?.(i, i + 1, e);
                  }
                }
                throw new FallbackExhaustedError(errors);
              }
              return inner!.next();
            },
          };
        },
      };
    },

    async callWithTools(
      request: Request,
      tools: ToolDef[]
    ): Promise<ToolResponse> {
      opts?.onRequest?.(request);
      const start = Date.now();
      const resp = await withFallback("callWithTools", (p) =>
        p.callWithTools!(request, tools)
      );
      opts?.onResponse?.(
        { content: resp.content, role: "assistant" },
        Date.now() - start
      );
      return resp;
    },

    capabilities(): string[] {
      const caps: string[] = [];
      if (primary.complete) caps.push("complete");
      if (primary.stream) caps.push("stream");
      if (primary.callWithTools) caps.push("callWithTools");
      return caps;
    },

    provider(): Provider {
      return primary;
    },

    close(): void {
      for (const p of resolvedProviders.values()) {
        p.close();
      }
    },
  };
}

File: sdk/py/hop_top_kit/llm.py
"""
hop_top_kit.llm — LLM provider abstraction layer.

Mirrors the Go ``llm`` package interfaces as Python Protocol classes,
provides a client facade via :func:`create_llm`, URI-based provider
resolution, YAML/env-var configuration, and a pluggable registry with
fallback support.

Provider adapters (e.g. LiteLLM) are registered as submodules via
entry-point plugins; this module contains only the core plumbing.
"""

from __future__ import annotations

import os
import time
from collections.abc import Callable, Iterator
from dataclasses import dataclass, field
from typing import Any, Protocol, runtime_checkable
from urllib.parse import parse_qs, urlparse

import yaml

from hop_top_kit.xdg import config_dir

__all__ = [  # noqa: RUF022 — grouped by category
    # Protocols
    "Completer",
    "Streamer",
    "ToolCaller",
    "Provider",
    # Data types
    "Message",
    "Request",
    "Response",
    "Usage",
    "Token",
    "ToolDef",
    "ToolCall",
    "ToolResponse",
    # URI + config
    "URI",
    "ProviderConfig",
    "ResolvedConfig",
    "parse_uri",
    "load_config",
    # Registry
    "register",
    "resolve",
    # Client
    "create_llm",
    "Client",
    # Router config
    "EvaConfig",
    "RouterConfig",
    "parse_router_config",
    "default_router_config",
    # Errors
    "LLMError",
    "ProviderNotFoundError",
    "CapabilityNotSupportedError",
    "AuthError",
    "RateLimitError",
    "ModelError",
    "FallbackExhaustedError",
    "RouterUnavailableError",
    "RouterTimeoutError",
    "ContractViolationError",
    "ThresholdInvalidError",
    "is_fallbackable",
]


# ---------------------------------------------------------------------------
# Protocols
# ---------------------------------------------------------------------------


@runtime_checkable
class Completer(Protocol):
    def complete(self, req: Request) -> Response: ...


@runtime_checkable
class Streamer(Protocol):
    def stream(self, req: Request) -> Iterator[Token]: ...


@runtime_checkable
class ToolCaller(Protocol):
    def call_with_tools(self, req: Request, tools: list[ToolDef]) -> ToolResponse: ...


class Provider(Protocol):
    def close(self) -> None: ...


# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------


@dataclass
class Message:
    role: str
    content: str


@dataclass
class Request:
    messages: list[Message]
    model: str = ""
    temperature: float = 0.0
    max_tokens: int = 0
    stop_sequences: list[str] = field(default_factory=list)
    extensions: dict[str, Any] = field(default_factory=dict)


@dataclass
class Response:
    content: str
    role: str = ""
    usage: Usage | None = None
    finish_reason: str = ""


@dataclass
class Usage:
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0


@dataclass
class Token:
    content: str = ""
    done: bool = False


@dataclass
class ToolDef:
    name: str
    description: str
    parameters: Any = None


@dataclass
class ToolCall:
    id: str
    name: str
    arguments: Any = None


@dataclass
class ToolResponse:
    content: str = ""
    tool_calls: list[ToolCall] = field(default_factory=list)


# ---------------------------------------------------------------------------
# URI + Config
# ---------------------------------------------------------------------------


@dataclass
class URI:
    scheme: str
    model: str
    host: str = ""
    params: dict[str, str] = field(default_factory=dict)


@dataclass
class ProviderConfig:
    api_key: str = ""
    base_url: str = ""
    model: str = ""
    params: dict[str, str] = field(default_factory=dict)
    extras: dict[str, Any] = field(default_factory=dict)


@dataclass
class ResolvedConfig:
    uri: URI
    provider: ProviderConfig
    fallbacks: list[str] = field(default_factory=list)


def parse_uri(raw: str) -> URI:
    """Parse a provider URI like ``openai://gpt-4?temperature=0.5``.

    Supports optional host (``ollama://localhost:11434/llama3``) and
    slash-containing model names (``openai://org/model``).
    """
    if not raw or "://" not in raw:
        raise LLMError(f"invalid LLM URI (missing scheme): {raw!r}")

    parsed = urlparse(raw)
    scheme = parsed.scheme
    if not scheme:
        raise LLMError(f"invalid LLM URI (empty scheme): {raw!r}")

    # Determine host vs model.
    # urlparse puts everything after :// into netloc when there's no path,
    # or splits netloc/path when a slash follows a host-like segment.
    host = ""
    model = ""

    try:
        port = parsed.port
    except ValueError:
        # Non-integer port (e.g. "mf:0.7") — not a real host:port.
        port = None

    # Heuristic: if there's a valid integer port, netloc is a real host.
    if port:
        hostname = parsed.hostname or ""
        host = f"{hostname}:{port}"
        model = parsed.path.lstrip("/")
    else:
        # No valid port — treat raw netloc + path as the model.
        # Use parsed.netloc (not parsed.hostname) to preserve
        # segments like "mf:0.7" that urlparse splits incorrectly.
        full = parsed.netloc or ""
        if parsed.path:
            full += parsed.path
        model = full.lstrip("/")

    # Query params — flatten single-value lists.
    params: dict[str, str] = {}
    for k, v_list in parse_qs(parsed.query).items():
        params[k] = v_list[0] if v_list else ""

    return URI(scheme=scheme, model=model, host=host, params=params)


def load_config(uri: str = "") -> ResolvedConfig:
    """Load LLM configuration from env vars, YAML file, or explicit *uri*.

    Resolution order (first wins per field):
    1. Explicit *uri* argument
    2. ``LLM_PROVIDER`` / ``LLM_API_KEY`` / ``LLM_BASE_URL`` / ``LLM_FALLBACK``
    3. ``{xdg.config_dir("hop")}/llm.yaml``
    """
    # Defaults from YAML file.
    yaml_data = _load_yaml_config()

    raw_uri = uri or os.environ.get("LLM_PROVIDER", "") or yaml_data.get("default", "")
    if not raw_uri:
        raise LLMError("no LLM provider configured (set LLM_PROVIDER env or provide a URI)")

    parsed = parse_uri(raw_uri)

    # Look up per-provider config from YAML ``providers`` dict.
    providers_map: dict[str, Any] = yaml_data.get("providers", {}) or {}
    prov_section = providers_map.get(parsed.scheme, {}) or {}

    api_key = os.environ.get("LLM_API_KEY", "") or prov_section.get("api_key", "")
    base_url = os.environ.get("LLM_BASE_URL", "") or prov_section.get("base_url", "")

    fallback_env = os.environ.get("LLM_FALLBACK", "")
    if fallback_env:
        fallbacks = [f.strip() for f in fallback_env.split(",") if f.strip()]
    else:
        fallbacks = yaml_data.get("fallback", []) or []

    provider_cfg = ProviderConfig(
        api_key=api_key,
        base_url=base_url,
        model=parsed.model,
        params=parsed.params,
    )

    return ResolvedConfig(uri=parsed, provider=provider_cfg, fallbacks=fallbacks)


def _load_yaml_config() -> dict[str, Any]:
    """Read ``{xdg.config_dir("hop")}/llm.yaml`` if it exists."""
    cfg_path = os.path.join(config_dir("hop"), "llm.yaml")
    if not os.path.isfile(cfg_path):
        return {}
    with open(cfg_path) as fh:
        data = yaml.safe_load(fh)
    return data if isinstance(data, dict) else {}


# ---------------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------------


class LLMError(Exception):
    """Base error for the LLM subsystem."""


class ProviderNotFoundError(LLMError):
    def __init__(self, scheme: str) -> None:
        self.scheme = scheme
        super().__init__(f"provider not found for scheme {scheme!r}")


class CapabilityNotSupportedError(LLMError):
    def __init__(self, capability: str, provider: str) -> None:
        self.capability = capability
        self.provider = provider
        super().__init__(f"capability {capability!r} not supported by {provider!r}")


class AuthError(LLMError):
    def __init__(self, provider: str) -> None:
        self.provider = provider
        super().__init__(f"authentication error for provider {provider!r}")


class RateLimitError(LLMError):
    def __init__(self, provider: str, retry_after: float = 0.0) -> None:
        self.provider = provider
        self.retry_after = retry_after
        super().__init__(f"rate limit hit for provider {provider!r} (retry after {retry_after}s)")


class ModelError(LLMError):
    def __init__(self, model: str, provider: str) -> None:
        self.model = model
        self.provider = provider
        super().__init__(f"model {model!r} not available on provider {provider!r}")


class FallbackExhaustedError(LLMError):
    def __init__(self, errors: list[Exception]) -> None:
        self.errors = errors
        super().__init__(f"all providers exhausted ({len(errors)} errors)")


class RouterUnavailableError(LLMError):
    def __init__(self, router: str, err: Exception | None = None) -> None:
        self.router = router
        self.err = err
        super().__init__(f"router {router!r} unavailable: {err}")


class RouterTimeoutError(LLMError):
    def __init__(self, router: str, timeout: float) -> None:
        self.router = router
        self.timeout = timeout
        super().__init__(f"router {router!r} timed out after {timeout}s")


class ContractViolationError(LLMError):
    def __init__(self, contract: str, violations: list[str]) -> None:
        self.contract = contract
        self.violations = violations
        super().__init__(f"contract {contract!r} violated: {'; '.join(violations)}")


class ThresholdInvalidError(LLMError):
    def __init__(self, threshold: float) -> None:
        self.threshold = threshold
        super().__init__(f"threshold {threshold:.4f} out of [0,1] range")


def is_fallbackable(err: Exception) -> bool:
    """Return True if *err* should trigger a fallback attempt."""
    return isinstance(err, (RateLimitError, RouterUnavailableError, RouterTimeoutError))


# ---------------------------------------------------------------------------
# Registry
# ---------------------------------------------------------------------------

Factory = Callable[[ResolvedConfig], Any]

_registry: dict[str, Factory] = {}


def register(scheme: str, factory: Factory) -> None:
    """Register a provider factory for *scheme*.

    Raises :class:`LLMError` if *scheme* is already registered.
    """
    if scheme in _registry:
        raise LLMError(f"scheme {scheme!r} already registered")
    _registry[scheme] = factory


def resolve(uri: str) -> Any:
    """Resolve a URI to a provider instance via the registry."""
    cfg = load_config(uri)
    scheme = cfg.uri.scheme
    factory = _registry.get(scheme)
    if factory is None:
        raise ProviderNotFoundError(scheme)
    return factory(cfg)


# ---------------------------------------------------------------------------
# Client facade
# ---------------------------------------------------------------------------


class Client:
    """Unified LLM client with fallback, hooks, and capability detection."""

    def __init__(
        self,
        provider: Any,
        fallback_providers: list[Any] | None = None,
        *,
        on_request: Callable[[Request], None] | None = None,
        on_response: Callable[[Response, float], None] | None = None,
        on_error: Callable[[Exception], None] | None = None,
        on_fallback: Callable[[int, int, Exception], None] | None = None,
        on_route: Callable[[str, float, str], None] | None = None,
        on_eva_result: Callable[[str, bool, list[str]], None] | None = None,
    ) -> None:
        self._provider = provider
        self._fallbacks = fallback_providers or []
        self._on_request = on_request
        self._on_response = on_response
        self._on_error = on_error
        self._on_fallback = on_fallback
        self._on_route = on_route
        self._on_eva_result = on_eva_result

    @property
    def provider(self) -> Any:
        """The primary provider instance."""
        return self._provider

    def capabilities(self) -> list[str]:
        """Return capability names supported by the primary provider."""
        caps: list[str] = []
        if isinstance(self._provider, Completer):
            caps.append("complete")
        if isinstance(self._provider, Streamer):
            caps.append("stream")
        if isinstance(self._provider, ToolCaller):
            caps.append("call_with_tools")
        return caps

    def complete(self, req: Request) -> Response:
        """Complete using primary provider, falling back on failure."""
        if self._on_request:
            self._on_request(req)

        providers = [self._provider, *self._fallbacks]
        errors: list[Exception] = []

        for i, prov in enumerate(providers):
            if not isinstance(prov, Completer):
                exc = CapabilityNotSupportedError("complete", type(prov).__name__)
                errors.append(exc)
                if i < len(providers) - 1 and self._on_fallback:
                    self._on_fallback(i, i + 1, exc)
                continue
            try:
                t0 = time.monotonic()
                resp = prov.complete(req)
                elapsed = time.monotonic() - t0
                if self._on_response:
                    self._on_response(resp, elapsed)
                return resp
            except Exception as exc:
                errors.append(exc)
                if self._on_error:
                    self._on_error(exc)
                if not is_fallbackable(exc):
                    raise
                # Fire fallback hook before trying next.
                if i < len(providers) - 1 and self._on_fallback:
                    self._on_fallback(i, i + 1, exc)
                continue

        raise FallbackExhaustedError(errors)

    def stream(self, req: Request) -> Iterator[Token]:
        """Stream tokens, falling back if provider lacks capability."""
        providers = [self._provider, *self._fallbacks]
        errors: list[Exception] = []

        for i, prov in enumerate(providers):
            if not isinstance(prov, Streamer):
                exc = CapabilityNotSupportedError("stream", type(prov).__name__)
                errors.append(exc)
                if i < len(providers) - 1 and self._on_fallback:
                    self._on_fallback(i, i + 1, exc)
                continue
            return prov.stream(req)

        raise CapabilityNotSupportedError("stream", type(self._provider).__name__)

    def call_with_tools(self, req: Request, tools: list[ToolDef]) -> ToolResponse:
        """Call with tools, falling back if provider lacks capability."""
        providers = [self._provider, *self._fallbacks]
        errors: list[Exception] = []

        for i, prov in enumerate(providers):
            if not isinstance(prov, ToolCaller):
                exc = CapabilityNotSupportedError("call_with_tools", type(prov).__name__)
                errors.append(exc)
                if i < len(providers) - 1 and self._on_fallback:
                    self._on_fallback(i, i + 1, exc)
                continue
            return prov.call_with_tools(req, tools)

        raise CapabilityNotSupportedError("call_with_tools", type(self._provider).__name__)

    def close(self) -> None:
        """Close primary and fallback providers."""
        if hasattr(self._provider, "close"):
            self._provider.close()
        for fb in self._fallbacks:
            if hasattr(fb, "close"):
                fb.close()


# ---------------------------------------------------------------------------
# Router config
# ---------------------------------------------------------------------------


@dataclass
class EvaConfig:
    """Evaluation/contract enforcement settings for RouteLLM."""

    contracts: list[str] = field(default_factory=list)
    enforce: bool = False


@dataclass
class RouterConfig:
    """RouteLLM router configuration extracted from provider extras."""

    base_url: str = "http://localhost:6060"
    grpc_port: int = 6061
    strong_model: str = ""
    weak_model: str = ""
    routers: list[str] = field(default_factory=list)
    router_config: dict[str, Any] = field(default_factory=dict)
    eva: EvaConfig = field(default_factory=EvaConfig)
    autostart: bool = False
    pid_file: str = ""


def default_router_config() -> RouterConfig:
    """Return a :class:`RouterConfig` with sensible defaults."""
    return RouterConfig()


def parse_router_config(extras: dict[str, Any]) -> RouterConfig:
    """Extract RouteLLM config from a provider's *extras* map.

    Looks for a ``routellm`` key in *extras*. Missing keys fall back
    to defaults. Environment variables override all other sources:

    - ``ROUTELLM_BASE_URL``
    - ``ROUTELLM_STRONG_MODEL``
    - ``ROUTELLM_WEAK_MODEL``
    - ``ROUTELLM_ROUTERS`` (comma-separated)
    """
    cfg = default_router_config()

    raw = extras.get("routellm")
    if raw is not None:
        if not isinstance(raw, dict):
            raise LLMError(f"routellm: expected dict, got {type(raw).__name__}")
        cfg.base_url = raw.get("base_url", cfg.base_url)
        cfg.grpc_port = int(raw.get("grpc_port", cfg.grpc_port))
        cfg.strong_model = raw.get("strong_model", cfg.strong_model)
        cfg.weak_model = raw.get("weak_model", cfg.weak_model)
        cfg.routers = raw.get("routers", cfg.routers)
        cfg.router_config = raw.get("router_config", cfg.router_config)
        cfg.autostart = bool(raw.get("autostart", cfg.autostart))
        cfg.pid_file = raw.get("pid_file", cfg.pid_file)

        eva_raw = raw.get("eva")
        if isinstance(eva_raw, dict):
            cfg.eva = EvaConfig(
                contracts=eva_raw.get("contracts", []),
                enforce=bool(eva_raw.get("enforce", False)),
            )

    # Environment variable overrides.
    if v := os.environ.get("ROUTELLM_BASE_URL"):
        cfg.base_url = v
    if v := os.environ.get("ROUTELLM_STRONG_MODEL"):
        cfg.strong_model = v
    if v := os.environ.get("ROUTELLM_WEAK_MODEL"):
        cfg.weak_model = v
    if v := os.environ.get("ROUTELLM_ROUTERS"):
        cfg.routers = [r.strip() for r in v.split(",") if r.strip()]

    return cfg


def create_llm(
    uri: str,
    *,
    fallback: list[str] | None = None,
    on_request: Callable[[Request], None] | None = None,
    on_response: Callable[[Response, float], None] | None = None,
    on_error: Callable[[Exception], None] | None = None,
    on_fallback: Callable[[int, int, Exception], None] | None = None,
    on_route: Callable[[str, float, str], None] | None = None,
    on_eva_result: Callable[[str, bool, list[str]], None] | None = None,
) -> Client:
    """Create a :class:`Client` from a provider URI with optional fallbacks.

    Args:
        uri: Provider URI (e.g. ``"openai://gpt-4"``).
        fallback: Optional list of fallback provider URIs.
        on_request: Hook called before each request.
        on_response: Hook called after a successful response.
        on_error: Hook called when a provider raises.
        on_fallback: Hook called when falling back ``(from_idx, to_idx, error)``.
        on_route: Hook called after a routing decision ``(router, score, model)``.
        on_eva_result: Hook called after eva evaluation ``(contract, passed, violations)``.

    Returns:
        A configured :class:`Client`.
    """
    cfg = load_config(uri)
    primary = _resolve_from_config(cfg)

    fallback_uris = cfg.fallbacks if fallback is None else fallback
    fallback_providers = []
    for fb_uri in fallback_uris:
        fb_cfg = load_config(fb_uri)
        fallback_providers.append(_resolve_from_config(fb_cfg))

    return Client(
        primary,
        fallback_providers,
        on_request=on_request,
        on_response=on_response,
        on_error=on_error,
        on_fallback=on_fallback,
        on_route=on_route,
        on_eva_result=on_eva_result,
    )


def _resolve_from_config(cfg: ResolvedConfig) -> Any:
    """Resolve a provider from a :class:`ResolvedConfig`."""
    factory = _registry.get(cfg.uri.scheme)
    if factory is None:
        raise ProviderNotFoundError(cfg.uri.scheme)
    return factory(cfg)
